event-queue and single-processing improvements

dkochmanski opened the pull request:
Fixes #586. More input-related changes PR will follow after this PR.

nsrahmad commented:
@dkochmanski In single-threaded mode when I run a demo from `demodemo` and close it, I get below condition. It doesn't happen when closing `demodemo` itself:
```lisp
There is no applicable method for the generic function
  #<STANDARD-GENERIC-FUNCTION CLIM:PROCESS-NEXT-EVENT (1)>
when called with arguments
  (NIL :TIMEOUT NIL).
   [Condition of type SB-PCL::NO-APPLICABLE-METHOD-ERROR]
See also:
  Common Lisp Hyperspec, 7.6.6 [:section]

Restarts:
 0: [RETRY] Retry calling the generic function.
 1: [ABORT] Return to application command loop.
 2: [RETRY] Retry SLIME REPL evaluation request.
 3: [*ABORT] Return to SLIME's top level.
 4: [REMOVE-FD-HANDLER] Remove #<SB-IMPL::HANDLER INPUT on descriptor 4: #<CLOSURE (LABELS SWANK/SBCL::RUN :IN SWANK/BACKEND:ADD-FD-HANDLER) {100486291B}>>
 5: [ABORT] Exit debugger, returning to top level.

Backtrace:
  0: ((:METHOD NO-APPLICABLE-METHOD (T)) #<STANDARD-GENERIC-FUNCTION CLIM:PROCESS-NEXT-EVENT (1)> NIL :TIMEOUT NIL) [fast-method]
  1: ((LAMBDA (SB-PCL::.ARG0. SB-INT:&MORE SB-PCL::.MORE-CONTEXT. SB-PCL::.MORE-COUNT.) :IN "/home/nsrahmad/quicklisp/setup.lisp") #<STANDARD-GENERIC-FUNCTION CLIM:PROCESS-NEXT-EVENT (1)> NIL :TIMEOUT NIL)
  2: (SB-PCL::CALL-NO-APPLICABLE-METHOD #<STANDARD-GENERIC-FUNCTION CLIM:PROCESS-NEXT-EVENT (1)> (NIL :TIMEOUT NIL))
  3: ((:METHOD CLIM-INTERNALS::EVENT-QUEUE-READ (CLIM-INTERNALS::SIMPLE-EVENT-QUEUE)) #<CLIM-INTERNALS::SIMPLE-EVENT-QUEUE {100B54BCE3}>) [fast-method]
  4: (CLIM-EXTENSIONS:SIMPLE-EVENT-LOOP #<CLIM-DEMO:DEMODEMO {100B54B833}>)
  5: ((:METHOD CLIM:READ-FRAME-COMMAND (CLIM:APPLICATION-FRAME)) #<CLIM-DEMO:DEMODEMO {100B54B833}> :STREAM NIL) [fast-method]
  6: ((FLET CALL-NEXT-METHOD :IN "/home/nsrahmad/common-lisp/McCLIM/Core/clim-core/frames.lisp"))
  7: ((:METHOD CLIM:READ-FRAME-COMMAND :AROUND (CLIM:APPLICATION-FRAME)) #<CLIM-DEMO:DEMODEMO {100B54B833}> :STREAM NIL) [fast-method]
  8: ((:METHOD CLIM:DEFAULT-FRAME-TOP-LEVEL (CLIM:APPLICATION-FRAME)) #<CLIM-DEMO:DEMODEMO {100B54B833}> :COMMAND-PARSER CLIM:COMMAND-LINE-COMMAND-PARSER :COMMAND-UNPARSER CLIM:COMMAND-LINE-COMMAND-UNPARSE..
  9: ((SB-PCL::EMF CLIM:DEFAULT-FRAME-TOP-LEVEL) #<unused argument> #<unused argument> #<CLIM-DEMO:DEMODEMO {100B54B833}>)
 10: ((LAMBDA (#:FRAME-ARG0) :IN "/home/nsrahmad/common-lisp/McCLIM/Examples/demodemo.lisp") #<CLIM-DEMO:DEMODEMO {100B54B833}>)
 11: ((FLET "FORM-FUN-23" :IN CLIM:RUN-FRAME-TOP-LEVEL))
 12: ((:METHOD CLIM:RUN-FRAME-TOP-LEVEL (CLIM:APPLICATION-FRAME)) #<CLIM-DEMO:DEMODEMO {100B54B833}>) [fast-method]
 13: ((FLET CALL-NEXT-METHOD :IN "/home/nsrahmad/common-lisp/McCLIM/Core/clim-core/frames.lisp"))
 14: ((FLET "FORM-FUN-53" :IN CLIM:RUN-FRAME-TOP-LEVEL))
 15: ((:METHOD CLIM:RUN-FRAME-TOP-LEVEL :AROUND (CLIM:APPLICATION-FRAME)) #<CLIM-DEMO:DEMODEMO {100B54B833}>) [fast-method]
 16: (CLIM-DEMO:DEMODEMO)
 17: (SB-INT:SIMPLE-EVAL-IN-LEXENV (CLIM-DEMO:DEMODEMO) #<NULL-LEXENV>)
 18: (EVAL (CLIM-DEMO:DEMODEMO))
 --more--
```

dkochmanski commented:
I can't reproduce this problem (probably because my display is too fast(?)). Should I put some sleeps in the code and try harder, or do you have a better idea? Also I don't think I've touched anything what could break it in this way. Are you certain it is a regression?

I've pushed a fix for the issue with closing frames. See commit message for a brief explanation what was going on.

dkochmanski commented:
good question. we could change it to utilize (buried for years) `schedule-event` extension! hang on, I'm testing McCLIM on ECL right now, I'll improve the demo afterwards.

dkochmanski commented:
I've updated graph-toy demo and fixed some remaining issues in event-queue vs schedule-event modules (deadlock quirk). I've also added fixes being part of my stashed input refactor and cleaned up scheduling module a bit more.

dkochmanski commented:
we are not prepending scheduled events in the queue, that causes weird behavior when peek/read are used one after another (peeked element could be different than the read one). pushed the fix squashed with previous commit.

nsrahmad commented:
with SBCL 1.4.14, when I close `graph-toy`, I get:
```lisp
failed AVER:
    (EQ SB-THREAD::ME (SB-THREAD::MUTEX-%OWNER SB-THREAD:MUTEX))
This is probably a bug in SBCL itself. (Alternatively, SBCL
might have been corrupted by bad user code, e.g. by an undefined
Lisp operation like (FMAKUNBOUND 'COMPILE), or by stray pointers
from alien code or from unsafe Lisp code; or there might be a
bug in the OS or hardware that SBCL is running on.) If it seems
to be a bug in SBCL itself, the maintainers would like to know
about it. Bug reports are welcome on the SBCL mailing lists,
which you can find at <http://sbcl.sourceforge.net/>.
   [Condition of type SB-INT:BUG]

Restarts:
 0: [ABORT] Return to application command loop.
 1: [RETRY] Retry SLIME REPL evaluation request.
 2: [*ABORT] Return to SLIME's top level.
 3: [ABORT] abort thread (#<THREAD "repl-thread" RUNNING {1003648603}>)

Backtrace:
  0: (SB-THREAD:CONDITION-WAIT #<SB-THREAD:WAITQUEUE Anonymous condition variable {1003E7BB33}> #<SB-THREAD:MUTEX "event queue" (free)> :TIMEOUT 0.0)
      Locals:
        #:.DEFAULTING-TEMP. = 0.0
        DEADLINEP = NIL
        MUTEX = #<SB-THREAD:MUTEX "event queue" (free)>
        QUEUE = #<SB-THREAD:WAITQUEUE Anonymous condition variable {1003E7BB33}>
        STOP-SEC = 508
        STOP-USEC = 558000
        TO-SEC = 0
        TO-USEC = 0
  1: (BORDEAUX-THREADS:CONDITION-WAIT #<SB-THREAD:WAITQUEUE Anonymous condition variable {1003E7BB33}> #<SB-THREAD:MUTEX "event queue" (free)> :TIMEOUT 0.0)
      Locals:
        #:.DEFAULTING-TEMP. = 0.0
        CONDITION-VARIABLE = #<SB-THREAD:WAITQUEUE Anonymous condition variable {1003E7BB33}>
        LOCK = #<SB-THREAD:MUTEX "event queue" (free)>
        TIMEOUT = 0.0
  2: (CLIM-SYS:CONDITION-WAIT #<SB-THREAD:WAITQUEUE Anonymous condition variable {1003E7BB33}> #<SB-THREAD:MUTEX "event queue" (free)> 0.0)
      Locals:
        CLIM-INTERNALS::CV = #<SB-THREAD:WAITQUEUE Anonymous condition variable {1003E7BB33}>
        CLIM-INTERNALS::LOCK = #<SB-THREAD:MUTEX "event queue" (free)>
        CLIM-INTERNALS::TIMEOUT = 0.0
  3: ((FLET SB-THREAD::WITH-RECURSIVE-LOCK-THUNK :IN CLIM-INTERNALS::EVENT-QUEUE-READ))
      Locals:
        CLIM-INTERNALS::CV = #<SB-THREAD:WAITQUEUE Anonymous condition variable {1003E7BB33}>
        CLIM-INTERNALS::LOCK = #<SB-THREAD:MUTEX "event queue" (free)>
        CLIM-INTERNALS::QUEUE = #<CLIM-INTERNALS::CONCURRENT-EVENT-QUEUE {1003EB35B3}>
        CLIM-INTERNALS::RESULT = NIL
        CLIM-INTERNALS::SCHEDULE-TIME = 508.558
        CLIM-INTERNALS::TIMEOUT = 0.0
  4: ((FLET "WITHOUT-INTERRUPTS-BODY-29" :IN SB-THREAD::CALL-WITH-RECURSIVE-LOCK))
      Locals:
        GOT-IT = T
        MUTEX = #<SB-THREAD:MUTEX "event queue" (free)>
  5: (SB-THREAD::CALL-WITH-RECURSIVE-LOCK #<CLOSURE (FLET SB-THREAD::WITH-RECURSIVE-LOCK-THUNK :IN CLIM-INTERNALS::EVENT-QUEUE-READ) {7FA8FD6E42BB}> #<SB-THREAD:MUTEX "event queue" (free)> T NIL)
      Locals:
        FUNCTION = #<CLOSURE (FLET SB-THREAD::WITH-RECURSIVE-LOCK-THUNK :IN CLIM-INTERNALS::EVENT-QUEUE-READ) {7FA8FD6E42BB}>
        GOT-IT = T
        MUTEX = #<SB-THREAD:MUTEX "event queue" (free)>
        TIMEOUT = NIL
        WAITP = T
  6: ((:METHOD CLIM-INTERNALS::EVENT-QUEUE-READ (CLIM-INTERNALS::CONCURRENT-EVENT-QUEUE)) #<CLIM-INTERNALS::CONCURRENT-EVENT-QUEUE {1003EB35B3}>) [fast-method]
  7: (CLIM-EXTENSIONS:SIMPLE-EVENT-LOOP #<CLIM-DEMO:DEMODEMO {100301FA43}>)
  8: ((:METHOD CLIM:READ-FRAME-COMMAND (CLIM:APPLICATION-FRAME)) #<CLIM-DEMO:DEMODEMO {100301FA43}> :STREAM NIL) [fast-method]
  9: ((FLET CALL-NEXT-METHOD :IN "/home/nsrahmad/common-lisp/McCLIM/Core/clim-core/frames.lisp"))
 10: ((:METHOD CLIM:READ-FRAME-COMMAND :AROUND (CLIM:APPLICATION-FRAME)) #<CLIM-DEMO:DEMODEMO {100301FA43}> :STREAM NIL) [fast-method]
 11: ((:METHOD CLIM:DEFAULT-FRAME-TOP-LEVEL (CLIM:APPLICATION-FRAME)) #<CLIM-DEMO:DEMODEMO {100301FA43}> :COMMAND-PARSER CLIM:COMMAND-LINE-COMMAND-PARSER :COMMAND-UNPARSER CLIM:COMMAND-LINE-COMMAND-UNPARSE..
 12: ((SB-PCL::EMF CLIM:DEFAULT-FRAME-TOP-LEVEL) #<unused argument> #<unused argument> #<CLIM-DEMO:DEMODEMO {100301FA43}>)
 13: ((LAMBDA (#:FRAME-ARG0) :IN "/home/nsrahmad/common-lisp/McCLIM/Examples/demodemo.lisp") #<CLIM-DEMO:DEMODEMO {100301FA43}>)
 14: ((FLET "FORM-FUN-23" :IN CLIM:RUN-FRAME-TOP-LEVEL))
 15: ((:METHOD CLIM:RUN-FRAME-TOP-LEVEL (CLIM:APPLICATION-FRAME)) #<CLIM-DEMO:DEMODEMO {100301FA43}>) [fast-method]
 16: ((FLET CALL-NEXT-METHOD :IN "/home/nsrahmad/common-lisp/McCLIM/Core/clim-core/frames.lisp"))
 17: ((FLET "FORM-FUN-53" :IN CLIM:RUN-FRAME-TOP-LEVEL))
 18: ((:METHOD CLIM:RUN-FRAME-TOP-LEVEL :AROUND (CLIM:APPLICATION-FRAME)) #<CLIM-DEMO:DEMODEMO {100301FA43}>) [fast-method]
 19: (CLIM-DEMO:DEMODEMO)
 --more--
```
For better error reporting, I switched to CCL which produces:
```lisp
? (clim-demo:demodemo)
> Error: There is no applicable method for the generic function:
>          #<STANDARD-GENERIC-FUNCTION CLIM-DEMO::VAL-ARRAY #x30200599ABEF>
>        when called with arguments:
>          (#<CLIM-DEMO:DEMODEMO #x302005CB3A7D>)
> While executing: #<CCL::STANDARD-KERNEL-METHOD NO-APPLICABLE-METHOD (T)>, in process listener(1).
> Type :GO to continue, :POP to abort, :R for a list of available restarts.
> If continued: Try calling it again
> Type :? for other options.
1 > :B
 (7F7CF96A6D20) : 0 (FUNCALL #'#<#<CCL::STANDARD-KERNEL-METHOD NO-APPLICABLE-METHOD (T)>> #<STANDARD-GENERIC-FUNCTION CLIM-DEMO::VAL-ARRAY #x30200599ABEF> #<CLIM-DEMO:DEMODEMO #x302005CB3A7D>) 149
 (7F7CF96A6D60) : 1 (DISPLAY-MAIN #<CLIM-DEMO::GRAPH-TOY #x302005EA6DAD> #<CLX-179150-APPLICATION-PANE CLIM-DEMO::MAIN-DISPLAY #x302005EAC13D>) 69
 (7F7CF96A6DA0) : 2 (%CALL-NEXT-METHOD ((#) #<STANDARD-METHOD CLIM:REDISPLAY-FRAME-PANE #> #<CLIM-DEMO::GRAPH-TOY #x302005EA6DAD> #<CLX-179150-APPLICATION-PANE CLIM-DEMO::MAIN-DISPLAY #x302005EAC13D> :FORCE-P ...)) 925
 (7F7CF96A6E20) : 3 (%%BEFORE-AND-AFTER-COMBINED-METHOD-DCODE ((#) #<STANDARD-METHOD CLIM:REDISPLAY-FRAME-PANE #> #<CLIM-DEMO::GRAPH-TOY #x302005EA6DAD> #<CLX-179150-APPLICATION-PANE CLIM-DEMO::MAIN-DISPLAY #x302005EAC13D> :FORCE-P ...)) 789
 (7F7CF96A6E90) : 4 (%CALL-NEXT-METHOD ((#) #<STANDARD-METHOD CLIM:REDISPLAY-FRAME-PANE #> #<CLIM-DEMO::GRAPH-TOY #x302005EA6DAD> #<CLX-179150-APPLICATION-PANE CLIM-DEMO::MAIN-DISPLAY #x302005EAC13D> :FORCE-P ...)) 509
 (7F7CF96A6F08) : 5 (FUNCALL #'#<#<STANDARD-METHOD CLIM:REDISPLAY-FRAME-PANE :AROUND (CLIM:APPLICATION-FRAME T)>> #<CLIM-DEMO::GRAPH-TOY #x302005EA6DAD> #<CLX-179150-APPLICATION-PANE CLIM-DEMO::MAIN-DISPLAY #x302005EAC13D> :FORCE-P NIL) 1149
 (7F7CF96A6FB0) : 6 (%%CNM-WITH-ARGS-COMBINED-METHOD-DCODE #(#<STANDARD-GENERIC-FUNCTION CLIM:REDISPLAY-FRAME-PANE #x3020011B6F7F> (#<#> #<#> #<#> #<#> #<#>) #(2 # #<STANDARD-GENERIC-FUNCTION CLIM:REDISPLAY-FRAME-PANE #x3020011B6F7F>) #<Compiled-function CCL::X-%%CHECK-KEYWORDS #x30000020FAFF> (#<#> NIL # #<#> #<#>)) 17521842146820 NIL) 1133
 (7F7CF96A7048) : 7 (%%CHECK-KEYWORDS #(2 #(:FORCE-P) #<Combined-Method CLIM:REDISPLAY-FRAME-PANE #x302005E24AFF>) 17521842146837) 229
 (7F7CF96A70D0) : 8 (%CALL-NEXT-METHOD ((NIL) #<STANDARD-METHOD CLIM:REDISPLAY-FRAME-PANE #> #<CLIM-DEMO::GRAPH-TOY #x302005EA6DAD> CLIM-DEMO::MAIN-DISPLAY)) 925
 (7F7CF96A7150) : 9 (FUNCALL #'#<#<STANDARD-METHOD CLIM:REDISPLAY-FRAME-PANE :AROUND (CLIM:APPLICATION-FRAME T)>> #<CLIM-DEMO::GRAPH-TOY #x302005EA6DAD> CLIM-DEMO::MAIN-DISPLAY :FORCE-P NIL) 1149
 (7F7CF96A71F8) : 10 (%%STANDARD-COMBINED-METHOD-DCODE (#<STANDARD-METHOD CLIM:REDISPLAY-FRAME-PANE :AROUND #> #<STANDARD-METHOD CLIM:REDISPLAY-FRAME-PANE #>) 17521842146891) 765
 (7F7CF96A7270) : 11 (%%CHECK-KEYWORDS #(2 #(:FORCE-P) #<Combined-Method #<Combined-Method CLIM:REDISPLAY-FRAME-PANE #x302005EB725F> #x302005EB8B0F>) 17521842146905) 229
 (7F7CF96A72E0) : 12 (FUNCALL #'#<#<STANDARD-METHOD CLIM:HANDLE-EVENT (CLIM-DEMO::GRAPH-TOY CLIM-DEMO::NEW-VALUE-EVENT)>> #<CLIM-DEMO::GRAPH-TOY #x302005EA6DAD> #<CLIM-DEMO::NEW-VALUE-EVENT #x302005F1288D>) 1101
 (7F7CF96A7328) : 13 (SIMPLE-EVENT-LOOP #<CLIM-DEMO:DEMODEMO #x302005CB3A7D>) 221
 (7F7CF96A7360) : 14 (%CALL-NEXT-METHOD (NIL #<STANDARD-METHOD CLIM:READ-FRAME-COMMAND #> . 17521842146976)) 1061
 (7F7CF96A73E0) : 15 (FUNCALL #'#<#<STANDARD-METHOD CLIM:READ-FRAME-COMMAND :AROUND (CLIM:APPLICATION-FRAME)>> #<CLIM-DEMO:DEMODEMO #x302005CB3A7D> :STREAM NIL) 773
 (7F7CF96A74A0) : 16 (%%STANDARD-COMBINED-METHOD-DCODE (#<STANDARD-METHOD CLIM:READ-FRAME-COMMAND :AROUND #> #<STANDARD-METHOD CLIM:READ-FRAME-COMMAND #>) 17521842146976) 765
 (7F7CF96A7520) : 17 (%%CHECK-KEYWORDS #(1 #(:STREAM) #<Combined-Method CLIM:READ-FRAME-COMMAND #x302005DDE11F>) 17521842146991) 229
 (7F7CF96A7598) : 18 (FUNCALL #'#<#<STANDARD-METHOD CLIM:DEFAULT-FRAME-TOP-LEVEL (CLIM:APPLICATION-FRAME)>> #<CLIM-DEMO:DEMODEMO #x302005CB3A7D> :COMMAND-PARSER CLIM:COMMAND-LINE-COMMAND-PARSER :COMMAND-UNPARSER CLIM:COMMAND-LINE-COMMAND-UNPARSER :PARTIAL-COMMAND-PARSER CLIM:COMMAND-LINE-READ-REMAINING-ARGUMENTS-FOR-PARTIAL-COMMAND :PROMPT "Command: ") 1661
 (7F7CF96A76F0) : 19 (%%CHECK-KEYWORDS #(1 #(:PROMPT :PARTIAL-COMMAND-PARSER :COMMAND-UNPARSER :COMMAND-PARSER) #<METHOD-FUNCTION CLIM:DEFAULT-FRAME-TOP-LEVEL #>) 17521842147050) 229
 (7F7CF96A7760) : 20 (FUNCALL #'#<#<STANDARD-METHOD CLIM:RUN-FRAME-TOP-LEVEL (CLIM:APPLICATION-FRAME)>> #<CLIM-DEMO:DEMODEMO #x302005CB3A7D>) 605
 (7F7CF96A77F0) : 21 (%CALL-NEXT-METHOD (NIL #<STANDARD-METHOD CLIM:RUN-FRAME-TOP-LEVEL #> . 17521842147140)) 1061
 (7F7CF96A7870) : 22 (FUNCALL #'#<#<STANDARD-METHOD CLIM:RUN-FRAME-TOP-LEVEL :AROUND (CLIM:APPLICATION-FRAME)>> #<CLIM-DEMO:DEMODEMO #x302005CB3A7D>) 1453
 (7F7CF96A79C0) : 23 (%%STANDARD-COMBINED-METHOD-DCODE (#<STANDARD-METHOD CLIM:RUN-FRAME-TOP-LEVEL :AROUND #> #<STANDARD-METHOD CLIM:RUN-FRAME-TOP-LEVEL #>) 17521842147140) 765
 (7F7CF96A7A30) : 24 (CALL-CHECK-REGS CLIM-DEMO:DEMODEMO) 221
 (7F7CF96A7A68) : 25 (TOPLEVEL-EVAL (CLIM-DEMO:DEMODEMO) NIL) 701
 (7F7CF96A7AF8) : 26 (READ-LOOP :INPUT-STREAM #<SYNONYM-STREAM to *TERMINAL-IO* #x3020003BAF7D> :OUTPUT-STREAM #<SYNONYM-STREAM to *TERMINAL-IO* #x3020003BAE1D> :BREAK-LEVEL 0 :PROMPT-FUNCTION #<Compiled-function (:INTERNAL CCL::READ-LOOP) (Non-Global)  #x30000054EB2F>) 2317
 (7F7CF96A7D58) : 27 (RUN-READ-LOOP :BREAK-LEVEL 0) 157
 (7F7CF96A7D80) : 28 (TOPLEVEL-LOOP) 101
 (7F7CF96A7DA8) : 29 (FUNCALL #'#<(:INTERNAL (TOPLEVEL-FUNCTION (CCL::LISP-DEVELOPMENT-SYSTEM T)))>) 101
 (7F7CF96A7DC8) : 30 (FUNCALL #'#<(:INTERNAL CCL::MAKE-MCL-LISTENER-PROCESS)>) 661
 (7F7CF96A7E60) : 31 (RUN-PROCESS-INITIAL-FORM #<TTY-LISTENER listener(1) [Active] #x3020003B9D7D> (#<COMPILED-LEXICAL-CLOSURE # #x3020003B98AF>)) 709
 (7F7CF96A7EF0) : 32 (FUNCALL #'#<(:INTERNAL (CCL::%PROCESS-PRESET-INTERNAL (PROCESS)))> #<TTY-LISTENER listener(1) [Active] #x3020003B9D7D> (#<COMPILED-LEXICAL-CLOSURE # #x3020003B98AF>)) 573
```
If I try to ignore and continue, I get:
```lisp
 (7F7CF96A7F98) : 33 (FUNCALL #'#<(:INTERNAL CCL::THREAD-MAKE-STARTUP-FUNCTION)>) 277
1 > :POP
> Error: The value -63.33278 is not of the expected type (REAL 0 9223372036854775807).
> While executing: CCL::MILLISECONDS, in process listener(1).
> Type :POP to abort, :R for a list of available restarts.
> Type :? for other options.
1 > :B  
 (7F7CF96A7198) : 0 (MILLISECONDS -63.33278) 109
 (7F7CF96A71C0) : 1 (%TIMED-WAIT-ON-SEMAPHORE-PTR #<A Foreign Pointer [gcable] #x7F7CF4062240> -63.33278 NIL) 445
 (7F7CF96A7250) : 2 (CONDITION-WAIT #<SEMAPHORE #x302005CB2DBD> #<RECURSIVE-LOCK "event queue" [ptr @ #x7F7CF40621C0] #x302005CB2E1D> :TIMEOUT -63.33278) 221
 (7F7CF96A72B0) : 3 (FUNCALL #'#<#<STANDARD-METHOD CLIM-INTERNALS::EVENT-QUEUE-READ (CLIM-INTERNALS::CONCURRENT-EVENT-QUEUE)>> #<CLIM-INTERNALS::CONCURRENT-EVENT-QUEUE #x302005CB30DD>) 605
 (7F7CF96A7328) : 4 (SIMPLE-EVENT-LOOP #<CLIM-DEMO:DEMODEMO #x302005CB3A7D>) 165
 (7F7CF96A7360) : 5 (%CALL-NEXT-METHOD (NIL #<STANDARD-METHOD CLIM:READ-FRAME-COMMAND #> . 17521842146976)) 1061
 (7F7CF96A73E0) : 6 (FUNCALL #'#<#<STANDARD-METHOD CLIM:READ-FRAME-COMMAND :AROUND (CLIM:APPLICATION-FRAME)>> #<CLIM-DEMO:DEMODEMO #x302005CB3A7D> :STREAM NIL) 773
 (7F7CF96A74A0) : 7 (%%STANDARD-COMBINED-METHOD-DCODE (#<STANDARD-METHOD CLIM:READ-FRAME-COMMAND :AROUND #> #<STANDARD-METHOD CLIM:READ-FRAME-COMMAND #>) 17521842146976) 765
 (7F7CF96A7520) : 8 (%%CHECK-KEYWORDS #(1 #(:STREAM) #<Combined-Method CLIM:READ-FRAME-COMMAND #x302005DDE11F>) 17521842146991) 229
 (7F7CF96A7598) : 9 (FUNCALL #'#<#<STANDARD-METHOD CLIM:DEFAULT-FRAME-TOP-LEVEL (CLIM:APPLICATION-FRAME)>> #<CLIM-DEMO:DEMODEMO #x302005CB3A7D> :COMMAND-PARSER CLIM:COMMAND-LINE-COMMAND-PARSER :COMMAND-UNPARSER CLIM:COMMAND-LINE-COMMAND-UNPARSER :PARTIAL-COMMAND-PARSER CLIM:COMMAND-LINE-READ-REMAINING-ARGUMENTS-FOR-PARTIAL-COMMAND :PROMPT "Command: ") 1661
 (7F7CF96A76F0) : 10 (%%CHECK-KEYWORDS #(1 #(:PROMPT :PARTIAL-COMMAND-PARSER :COMMAND-UNPARSER :COMMAND-PARSER) #<METHOD-FUNCTION CLIM:DEFAULT-FRAME-TOP-LEVEL #>) 17521842147050) 229
 (7F7CF96A7760) : 11 (FUNCALL #'#<#<STANDARD-METHOD CLIM:RUN-FRAME-TOP-LEVEL (CLIM:APPLICATION-FRAME)>> #<CLIM-DEMO:DEMODEMO #x302005CB3A7D>) 605
 (7F7CF96A77F0) : 12 (%CALL-NEXT-METHOD (NIL #<STANDARD-METHOD CLIM:RUN-FRAME-TOP-LEVEL #> . 17521842147140)) 1061
 (7F7CF96A7870) : 13 (FUNCALL #'#<#<STANDARD-METHOD CLIM:RUN-FRAME-TOP-LEVEL :AROUND (CLIM:APPLICATION-FRAME)>> #<CLIM-DEMO:DEMODEMO #x302005CB3A7D>) 1453
 (7F7CF96A79C0) : 14 (%%STANDARD-COMBINED-METHOD-DCODE (#<STANDARD-METHOD CLIM:RUN-FRAME-TOP-LEVEL :AROUND #> #<STANDARD-METHOD CLIM:RUN-FRAME-TOP-LEVEL #>) 17521842147140) 765
 (7F7CF96A7A30) : 15 (CALL-CHECK-REGS CLIM-DEMO:DEMODEMO) 221
 (7F7CF96A7A68) : 16 (TOPLEVEL-EVAL (CLIM-DEMO:DEMODEMO) NIL) 701
 (7F7CF96A7AF8) : 17 (READ-LOOP :INPUT-STREAM #<SYNONYM-STREAM to *TERMINAL-IO* #x3020003BAF7D> :OUTPUT-STREAM #<SYNONYM-STREAM to *TERMINAL-IO* #x3020003BAE1D> :BREAK-LEVEL 0 :PROMPT-FUNCTION #<Compiled-function (:INTERNAL CCL::READ-LOOP) (Non-Global)  #x30000054EB2F>) 2317
 (7F7CF96A7D58) : 18 (RUN-READ-LOOP :BREAK-LEVEL 0) 157
 (7F7CF96A7D80) : 19 (TOPLEVEL-LOOP) 101
 (7F7CF96A7DA8) : 20 (FUNCALL #'#<(:INTERNAL (TOPLEVEL-FUNCTION (CCL::LISP-DEVELOPMENT-SYSTEM T)))>) 101
 (7F7CF96A7DC8) : 21 (FUNCALL #'#<(:INTERNAL CCL::MAKE-MCL-LISTENER-PROCESS)>) 661
 (7F7CF96A7E60) : 22 (RUN-PROCESS-INITIAL-FORM #<TTY-LISTENER listener(1) [Semaphore timed wait] #x3020003B9D7D> (#<COMPILED-LEXICAL-CLOSURE # #x3020003B98AF>)) 709
 (7F7CF96A7EF0) : 23 (FUNCALL #'#<(:INTERNAL (CCL::%PROCESS-PRESET-INTERNAL (PROCESS)))> #<TTY-LISTENER listener(1) [Semaphore timed wait] #x3020003B9D7D> (#<COMPILED-LEXICAL-CLOSURE # #x3020003B98AF>)) 573
 (7F7CF96A7F98) : 24 (FUNCALL #'#<(:INTERNAL CCL::THREAD-MAKE-STARTUP-FUNCTION)>) 277
1 > 
```

nsrahmad commented:
SBCL also shows above CCL's `no-applicable-method` or the crash depending on the timing of closing.

dkochmanski commented:
no-applicable-method was trivial to fix. as of aver problem, the gist of it may be summarized as follows:

    (sb-thread:condition-wait (sb-thread:make-waitqueue) (sb-thread:make-mutex))

above will trigger the very same aver. When we exit frame lock is released in unwind-protect. condition-wait still hangs for the next scheduled event (with timeout), and when it is notified it wakes up in the environment it is not very happy about. I'm still thinking about the fix.

dkochmanski commented:
and from more esoteric analysis:

```
(with-recursive-lock-held (lock)
      (loop
         with schedule-time = (event-schedule-time queue)
         as timeout = (when schedule-time (- schedule-time (now)))
         as result = (%event-queue-read queue)
         if result do (return result)
         else do
           (unless (sb-thread:mutex-value lock)
             (log:error "mutex ~s is free!" lock))
           (when timeout (log:error "condition-wait start (~s)" timeout))
           (condition-wait cv lock timeout)
           (when timeout (log:error "condition-wait done"))))
```

in this code it may happen, that mutex is free… so it is sbcl internal issue too

dkochmanski commented:
so the fix seems to be less exciting than I have envisioned, we simply need to acquire lock after each condition-wait if we want it on. This is documented sbcl behavior. Please see if it works OK now.

I've also tweaked a problem when frame is exited (disown-frame could be called without a frame manager what lead to an error). Frame exiting could use some more work still.

nsrahmad commented:
In `Pane Hierarchy viewer` demo, Scrolling the description text pane to bottom lands in debugger with:
```lisp
The value
  NIL
is not of type
  CHARACTER
when setting an element of (ARRAY CHARACTER)
   [Condition of type TYPE-ERROR]

Restarts:
 0: [ABORT] Return to application command loop.
 1: [ABORT] Return to application command loop.
 2: [RETRY] Retry SLIME REPL evaluation request.
 3: [*ABORT] Return to SLIME's top level.
 4: [ABORT] abort thread (#<THREAD "repl-thread" RUNNING {1002808603}>)

Backtrace:
  0: ((SB-VM::OPTIMIZED-DATA-VECTOR-SET CHARACTER) #<unavailable argument> #<unavailable argument> #<unavailable argument>)
  1: (VECTOR-PUSH-EXTEND NIL "\0" NIL)
  2: (CLIM:COMPLETE-INPUT #<CLIM:STANDARD-INPUT-EDITING-STREAM {100291D803}> #<CLOSURE (LAMBDA (CLIM-INTERNALS::SO-FAR CLIM-INTERNALS::MODE) :IN CLIM-INTERNALS::%ACCEPT) {1002E3C32B}> :PARTIAL-COMPLETERS (..
  3: ((:METHOD CLIM-INTERNALS::%ACCEPT (CLIM-INTERNALS::|(presentation-type CLIM::COMMAND-NAME)| T T CLIM:TEXTUAL-VIEW)) #<CLIM-INTERNALS::|(presentation-type CLIM::COMMAND-NAME)| {1004665C43}> (CLIM:COMMA..
  4: ((SB-PCL::EMF CLIM-INTERNALS::%ACCEPT) #<unused argument> #<unused argument> #<CLIM-INTERNALS::|(presentation-type CLIM::COMMAND-NAME)| {1004665C43}> (CLIM:COMMAND-NAME :COMMAND-TABLE #<CLIM:STANDARD-..
  5: ((LAMBDA (SB-PCL::.ARG0. SB-PCL::.ARG1. SB-PCL::.ARG2. SB-PCL::.ARG3. SB-INT:&MORE SB-PCL::.MORE-CONTEXT. SB-PCL::.MORE-COUNT.)) #<CLIM-INTERNALS::|(presentation-type CLIM::COMMAND-NAME)| {1004665C43}..
  6: ((FLET "INPUT-CONT22" :IN CLIM:ACCEPT-1))
  7: (CLIM-INTERNALS::INVOKE-HANDLE-EMPTY-INPUT #<CLIM:STANDARD-INPUT-EDITING-STREAM {100291D803}> #<CLOSURE (FLET "INPUT-CONT22" :IN CLIM:ACCEPT-1) {7F15C6506ABB}> #<CLOSURE (FLET "HANDLER-CONT23" :IN CLI..
  8: ((LAMBDA (STREAM) :IN CLIM:ACCEPT-1) #<CLIM:STANDARD-INPUT-EDITING-STREAM {100291D803}>)
  9: ((:METHOD CLIM-INTERNALS::INVOKE-WITH-INPUT-EDITING (T T T T T)) #<CLIM:STANDARD-INPUT-EDITING-STREAM {100291D803}> #<CLOSURE (LAMBDA (STREAM) :IN CLIM:ACCEPT-1) {1002E3553B}> #<unused argument> #<unu..
 10: ((FLET CALL-NEXT-METHOD :IN "/home/nsrahmad/common-lisp/McCLIM/Core/clim-core/input-editing.lisp"))
 11: ((:METHOD CLIM-INTERNALS::INVOKE-WITH-INPUT-EDITING (CLIM:INPUT-EDITING-STREAM T T T T)) #<CLIM:STANDARD-INPUT-EDITING-STREAM {100291D803}> #<CLOSURE (LAMBDA (STREAM) :IN CLIM:ACCEPT-1) {1002E3553B}> ..
 12: ((FLET CALL-NEXT-METHOD :IN "/home/nsrahmad/common-lisp/McCLIM/Core/clim-core/input-editing.lisp"))
 13: ((:METHOD CLIM-INTERNALS::INVOKE-WITH-INPUT-EDITING :AROUND (T T T T T)) #<CLIM:STANDARD-INPUT-EDITING-STREAM {100291D803}> #<CLOSURE (LAMBDA (STREAM) :IN CLIM:ACCEPT-1) {1002E3553B}> #<CLOSURE (LAMBD..
 14: (CLIM:ACCEPT-1 #<CLIM:STANDARD-INPUT-EDITING-STREAM {100291D803}> (CLIM:COMMAND-NAME :COMMAND-TABLE #<CLIM:STANDARD-COMMAND-TABLE CLIM-DEMO::HIERARCHY {1001EB1CA3}>) :VIEW NIL :DEFAULT NIL :DEFAULT-TY..
 15: ((:METHOD CLIM:STREAM-ACCEPT (CLIM:STANDARD-INPUT-EDITING-STREAM T)) #<CLIM:STANDARD-INPUT-EDITING-STREAM {100291D803}> (CLIM:COMMAND-NAME :COMMAND-TABLE #<CLIM:STANDARD-COMMAND-TABLE CLIM-DEMO::HIERA..
 16: ((SB-PCL::EMF CLIM:STREAM-ACCEPT) #<unused argument> #<unused argument> #<CLIM:STANDARD-INPUT-EDITING-STREAM {100291D803}> (CLIM:COMMAND-NAME :COMMAND-TABLE #<CLIM:STANDARD-COMMAND-TABLE CLIM-DEMO::HI..
 17: ((LABELS CLIM-INTERNALS::DO-ACCEPT :IN CLIM:ACCEPT) (:HISTORY NIL :PROMPT NIL :HISTORY NIL))
 18: (CLIM:ACCEPT (CLIM:COMMAND-NAME :COMMAND-TABLE #<CLIM:STANDARD-COMMAND-TABLE CLIM-DEMO::HIERARCHY {1001EB1CA3}>) :STREAM #<CLIM:STANDARD-INPUT-EDITING-STREAM {100291D803}> :PROMPT NIL :HISTORY NIL)
 19: (CLIM:COMMAND-LINE-COMMAND-PARSER #<CLIM:STANDARD-COMMAND-TABLE CLIM-DEMO::HIERARCHY {1001EB1CA3}> #<CLIM:STANDARD-INPUT-EDITING-STREAM {100291D803}>)
 --more--
```

nsrahmad commented:
@dkochmanski Same error with `Coordinate swizzling` demo.

dkochmanski commented:
I've made yet another overhaul of the implementation. It takes into account spurious locks and makes a clear difference between decay time and the real timeout (the last one was a regression which you have reported above).

Single processing is now flacky when used with schedule-event due to clx quirks (see commit). I'll investigate that tomorrow, so please do not merge it yet. If that can't be fixed on clx I'll make it a busy-wait instead of fine-grained timeouts.

dkochmanski commented:
OK, this should be fine as it is now. some changes improving ecl support wrt timeouts will land in clx soon, but the problem with timout is not related to mcclim whatsoever. on sbcl and ccl this should already work (for single-processed run as well).

nsrahmad commented:
Looks good, Merging it now.

dkochmanski commented:
Thank you for working with me on that. This contribution would be much worse without your help.
